compiler/hash-attrs/src/target.rs

```diff
@@ -0,0 +1,292 @@
+        match expr {
+            ast::Expr::ConstructorCall(_) => AttrTarget::ConstructorCall,
+            ast::Expr::Macro(_) => AttrTarget::MacroInvocation,
+            ast::Expr::Unsafe(_) => AttrTarget::Unsafe,
+            ast::Expr::Lit(_) => AttrTarget::Lit,
+            ast::Expr::Block(ast::BlockExpr { data: block }) => match block.body() {
+                ast::Block::Loop(_) | ast::Block::While(_) | ast::Block::For(_) => AttrTarget::Loop,
+                ast::Block::Match(_) | ast::Block::If(_) => AttrTarget::Match,
+                ast::Block::Body(_) => AttrTarget::Block,
+            },
+            ast::Expr::Import(_) => AttrTarget::Import,
+            ast::Expr::StructDef(_) => AttrTarget::StructDef,
+            ast::Expr::EnumDef(_) => AttrTarget::EnumDef,
+            ast::Expr::TyFnDef(_) => AttrTarget::TyFnDef,
+            ast::Expr::TraitDef(_) => AttrTarget::Item,
+            ast::Expr::ImplDef(_) => AttrTarget::ImplDef,
+            ast::Expr::ModDef(_) => AttrTarget::ModDef,
+            ast::Expr::FnDef(_) => AttrTarget::FnDef,
+            ast::Expr::Ty(_) => AttrTarget::Ty,
+
+            // If this is a declaration, we have to recurse into the subject...
+            ast::Expr::Declaration(ast::Declaration { value: Some(value), .. }) => {
+                AttrTarget::classify_expr(value.body())
+            }
+            _ => AttrTarget::Expr,
+        }
+    }
+}
+
+impl fmt::Display for AttrTarget {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let item_count = self.bits().count_ones();
+
+        // We will collect all of the allowed argument kinds into a vector
+        let mut allowed_argument_kinds = Vec::with_capacity(item_count as usize);
+
+        for kind in (*self).iter() {
+            match kind {
+                AttrTarget::ConstructorCall => allowed_argument_kinds.push("constructor call"),
+                AttrTarget::MacroInvocation => allowed_argument_kinds.push("directive"),
+                AttrTarget::Unsafe => allowed_argument_kinds.push("unsafe expression"),
+                AttrTarget::Lit => allowed_argument_kinds.push("literal"),
+                AttrTarget::Loop => allowed_argument_kinds.push("loop block"),
+                AttrTarget::Match => allowed_argument_kinds.push("match block"),
+                AttrTarget::ImplDef => allowed_argument_kinds.push("impl block"),
+                AttrTarget::Mod => allowed_argument_kinds.push("module"),
+                AttrTarget::ModDef => allowed_argument_kinds.push("mod block"),
+                AttrTarget::Block => allowed_argument_kinds.push("body block"),
+                AttrTarget::Import => allowed_argument_kinds.push("import"),
+                AttrTarget::StructDef => allowed_argument_kinds.push("`struct` definition"),
+                AttrTarget::EnumDef => allowed_argument_kinds.push("`enum` definition"),
+                AttrTarget::TyFnDef => allowed_argument_kinds.push("type function definition"),
```

Comment:
This is a bit unrelated to this PR, but type functions are called implicit functions in the TIR. Maybe we should unify these two names?